servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -260,15 +248,50 @@ public void onComplete() {
         final ThreadLocalRandom rnd = ThreadLocalRandom.current();
 
         // Try first to see if an existing connection can be used
-        final List<C> connections = host.connections;
-        final int size = connections.size();
-        // With small enough search space, attempt all connections.
-        // Back off after exploring most of the search space, it gives diminishing returns.
-        final int attempts = size < MIN_SEARCH_SPACE ? size : (int) (size * SEARCH_FACTOR);
-        for (int i = 0; i < attempts; i++) {
-            final C connection = connections.get(rnd.nextInt(size));
-            if (selector.test(connection)) {
-                return succeeded(connection);
+        final Object[] connections = host.connections;
+        if (connections.length <= 64) {
+            // With small enough search space, attempt all connections.
+            // Back off after exploring most of the search space, it gives diminishing returns.
+            int i = 0;
+            long indexMask = 0;
+            int collisions = 0;
+            final int collisionThreshold = connections.length >>> 1;
+            while (i < connections.length) {
+                int index = rnd.nextInt(connections.length);
```

Comment:
This algorithm is a bit involved and may have collisions. In the worst case it iterates over all connections. Consider taking a random value only for the first test and then continue in a loop over all connections.

Reply:
I will experiment with this, and I'll move the selection changes to another PR.